### ISSUE TITLE
Use a per customer configmap instead of a per customer workflow

### ISSUE THREAD

**Issue by tumido** [aicoe-aiops/sync-pipelines#5]
Currently we store the customer specific workflow in a configmap with prepopulated per-customer default parameters. This workflow is rendered from a template via Ansible.

In order to aicoe-aiops/sync-pipelines-private#7 we need to have those config values loaded from the workflow dynamically from a configmap, since we can't template in Kustomize so the current approach would require a huge copypasting when setting up a new pipeline -> maintenance hell.

**Comment by durandom** [comment]
How about reading the config from a file? I.e. removing the dependency on configmaps from solgate?

**Comment by tumido** [comment]
@durandom yes, you're thinking the same as I do, but I'm struggling to explain properly. :smile: .

Current design as outlined in https://github.com/aicoe-aiops/sync-pipelines/pull/9 expects a single configuration file per instance called `config.ini`. This file is mounted into the solgate container and read in runtime from a standard default location. No passing of environment variables, no extra values, no configmaps. This instance specific config file will be stored in OpenShift as a secret (naturally, since it contains credentials).

I want to have these config files tracked in the repo in as readable format as possible (so only the creds are encrypted, everything else is review-able on plain sight). And we can achieve that by using [kustomize-sopssecretgenerator](https://github.com/goabout/kustomize-sopssecretgenerator) (instead of KSOPS). With KSOPS we would have to track the config files as Kubernetes "Secret" resources (everything is base64 encoded + selected values are encrypted on top of that), so it can't be easily reviewed. With sopssecretgenerator, we can generate the secrets on the fly during `kustomize build` and keep them in plain text with only the cred values encrypted.

**Comment by durandom** [comment]
So you'll be mixing secrets for ENV variables and having the config unencrypted, like the order of reading the secrets here: https://github.com/thanos-io/thanos/blob/master/docs/storage.md#credentials

That way the secret doesnt appear in the config file in the first place?

**Comment by tumido** [comment]
Let's rather demonstrate it all by an example. :wink:

Unfortunately we can't use environment variables for the secrets, since each pipeline instance may use different number of sync locations, therefore different amount of credentials.

I want to store all instance configuration in a single config file. This config is stored as a secret in OpenShift and [mounted as a file in runtime](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod) and looks like this:

```ini
; /etc/solgate.ini

[source_raw]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
formatter = {date}/{collection}.{ext}
base_path = DH-PLAYPEN/storage/input

[destination_unpack_historic]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
formatter = {collection}/historic/{date}-{collection}.{ext}
base_path = DH-PLAYPEN/storage/output
unpack = yes

[destination_unpack_latest]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
formatter = {collection}/latest/full_data.csv
base_path = DH-PLAYPEN/storage/output
unpack = yes

[destination_raw]
aws_access_key_id = FVCOYKA2QS37NFEFFA58
aws_secret_access_key = chC9tYxarMOq3lYMpnFDNayDLTqgJ6O9Tbh0EQpZ
endpoint_url = https://s3.upshift.redhat.com
base_path = DH-PLAYPEN/storage/output

[solgate]
alerts_smtp_server = smtp.corp.redhat.com
alerts_from        = solgate-alerts@redhat.com
alerts_to          = dev-null@redhat.com
timedelta          = 1d
```

As you can see, this instance will sync from `source_raw` to 3 destinations `destination_*`

We want to track this INI file in GIT in as readible format as possible. So in GIT, we encrypt only `aws` specific keys of it via SOPS:
```yaml
# manifests/.sops.yaml

creation_rules:
  - encrypted_regex: "^(aws.*)$"
    pgp: "EFDB9AFBD18936D9AB6B2EECBD2C73FF891FBC7E, A76372D361282028A99F9A47590B857E0288997C, 04DAFCD9470A962A2F272984E5EB0DA32F3372AC"
```

And the resulting config file will look like this:

```ini
; manifests/overlays/dev/this_instance/config.ini
[source]
aws_access_key_id     = ENC[AES256_GCM,data:lpy0WhN4e+o5spYSyBr4G0Bp5W8=,iv:9G/7iI/E46vU5k+gxcMHW1mmArzeqHYr76a4Lg8q2/4=,tag:gAdwPCSx9axGwojRyEQQWA==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:PZ91CnbyWNHX9aTVSNR0HDWTTMPMtaGMxWGYshKH1M7Vo5xvCh1QHw==,iv:/+jUkv7TLRdrRVzu8DG8tC05snT5mlaT6XAq+lslS4k=,tag:e4SjZGSzlnm5LMyej9U5hQ==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
formatter             = {date}/{collection}.{ext}
base_path             = DH-PLAYPEN/storage/input

[destination_unpack_historic]
aws_access_key_id     = ENC[AES256_GCM,data:96ftEMbEpt4qAi/6LK4pzCyJLDM=,iv:U6dGaPBZXfee7XCj1ZhmGu75FJ9Z4mjDVtlVkKq4uDE=,tag:seihxPfpNB4EEbP1Wq3CDw==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:CGMHXL0abcviOXTe/gQ/dvegn2iXp3hwdPAYIXjnjMHq/Wcz0Kv43Q==,iv:IguBm4gWku76kRGtjrU7yb+LK749oWDh2R9sdMpytx8=,tag:JXqbKa/36ESV4tTkOFMYyQ==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
formatter             = {collection}/historic/{date}-{collection}.{ext}
base_path             = DH-PLAYPEN/storage/output
unpack                = yes

[destination_unpack_latest]
aws_access_key_id     = ENC[AES256_GCM,data:d3cntWCYKdjldnf+feRiNcFoL/s=,iv:SP158uN1Vat1P+RipnQBO8328Z6smuo91iPwco1720k=,tag:EWyNsngiv/sddqKAAk35nQ==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:8CzkDrXCKBHnkHA7iP4E7YsglBRd+kTRzSidBICgofPr6pu+eEYFIA==,iv:AEiL4jc0L7fsqfT/9vErsHSOkbAqjYuCFU05ngoKoNk=,tag:q6SnLrl5tiQUC2j0PBjOUg==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
formatter             = {collection}/latest/full_data.csv
base_path             = DH-PLAYPEN/storage/output
unpack                = yes

[destination_raw]
aws_access_key_id     = ENC[AES256_GCM,data:0eroBlaIxNzhCzbgf+VZensrlyA=,iv:yGiiLyIq4x2lhB7lMKuR0A/0btwQcuIDJTpnS80HD6Q=,tag:G4M+Lj5u0XqmMdCAaKJ4pA==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:ah+uxPlDQ0I6i7ZUagWpqno735GxEF8bqwKc2jZ0mNSVBuS9Xyj5uQ==,iv:myzLeh7goVyyoXdHP2IPvCcmsVo7wRJ+Bm3e56Apxk4=,tag:ne1oKcPVFK46PjS0x2KnKw==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
base_path             = DH-PLAYPEN/storage/output

[solgate]
alerts_smtp_server = smtp.corp.redhat.com
alerts_from        = solgate-alerts@redhat.com
alerts_to          = dev-null@redhat.com
timedelta          = 1d

[sops]
pgp__list_0__map_fp         = EFDB9AFBD18936D9AB6B2EECBD2C73FF891FBC7E
pgp__list_2__map_created_at = 2020-08-07T10:29:29Z
pgp__list_2__map_fp         = 04DAFCD9470A962A2F272984E5EB0DA32F3372AC
lastmodified                = 2020-08-07T10:29:30Z
pgp__list_1__map_enc        = -----BEGIN PGP MESSAGE-----\n\nhQIMA5MpUvZdp4z9AQ//WgCo6KcapKGj1sQopw579+UyCFL0zeEQwVElsmKKVg5N\nvnncjEATrSMo5Ojvmi5JsXka9S8pCx19nB1TuwNTbeCHr901e0W6yy6Qvoe901Ei\npJtGgZChckPpg8JozDU1paYlbS/YWd5hDbKukd6BGSiKViHhAyWtYVg1A/4kFY06\nCflgJFwvmGTUf6W6zkOYjpfK/MvMYi/8zpSo5myRNf7MZJmDStFRlp/f2+SxrnCV\njf3Qy00a6mJtYDLvTDlydhdjLXHA7J4NARah7lXqHM2s5gmeUtUB1DNtB9q3Th3N\npsfM+5xwOacATX7eIVxJ5PbjZ+c+ODVBH2Jzysw5YDvdnr4dGsUzngvAx0JqLCxf\nHy/g8Kxd8ZWx0yMyIY3Kfvi4a+lvZ+wiiW8CcLQCLWmcuQJ8dFWm3/P03Hf9Jgbj\n2K8hlf3PhBr7mgnzG5h9mYIHBRSwFYMZ05GgKn5mMMwtp6FvFrY5xRPO1yuGWRPJ\nKYts34EKITRZnvQB8/nK/V3H6C7mVFYGIvTFysH/XNS/zoVLNEfuKvxiSSsvGZM2\nCuLZBAgz7TMUg0x5Di3VCMCaSqYAaYwDqGoeQAZgdD+fvknhzdXdu45NwoabTCNZ\nOdjENyjrAUBwZFvO4w6qK1NWswcl2qdXMtxryE9s1kvpd8e+UaE0yom+a5v5LW/S\nXAGuE/4FhljCA/3IdLrKHV3NTMlMo4CNgLQHBdfBW5jR0NbSeyUtYczOsme/+c9J\n5mMOW9NW4WvBODEEjIj95AvjLWNP4s/FR0u6gQ925ZQhl0vCfbakY0bfnMaD\n=uaW0\n-----END PGP MESSAGE-----\n
mac                         = ENC[AES256_GCM,data:oKrctMbVjP4s8natZaKyMgERF58xzY3ANuWFZoJ+FMQHa9wdX+8auWAkG3T3+1hZ1Si3rzso18yAmWnu5XJJ5MHeFv151mYMfmPGuD/umxA1rs1O86AMxMD+F4WOrJV8Q+Oq6CvkUnvimoF43bkV5y5kEEtOP4m1Fo42bywBkyI=,iv:FwiWswP0czEObEZBXLYXrx3zY4sf32hktQChMR7mLyk=,tag:wb7EcS/BASG5MIZBwz/lOw==,type:str]
version                     = 3.6.0
pgp__list_0__map_created_at = 2020-08-07T10:29:29Z
pgp__list_1__map_created_at = 2020-08-07T10:29:29Z
pgp__list_1__map_fp         = A76372D361282028A99F9A47590B857E0288997C
encrypted_regex             = ^(aws.*)$
pgp__list_0__map_enc        = -----BEGIN PGP MESSAGE-----\n\nhQEMA/irrHa183bxAQf9HR9cJM+dhrmO5VkVwcj/Lh3+/CcclK0SDFVP3Hz0swH5\nUv04lQaoa9n/yap/633J61HnlWpitWE3HKnI+56fCaQb5nSZcRVU82mIVpK2UN0y\n5UwJq8tg5DzkzjFgl6IolYCeEB4mvqBgUD04hf6M65H2tzuwOOvvrknv6gUekorJ\nN4VRwzUZRZwiYFxhLRj/kcWxCd+bdOGge4EfkN2RoVNRFXkQ0CNawEzxg3e6lFk2\n6ouMs1Geu2aZs/LO0sNliNidW9p6uUUFNuJ7GWdfIjwOTmZSVIBzPBpU1nvsdyhB\npZBdRJHUoYt3mu1uxkf9J9hKfcBqcbulUXDS9qHxrtJcASZAlAMTONjmeZcVcPf5\n1RGT0uMp3L6lS22rt7FyAFnbaDJ1h8iUuPpHCX+8T67rXX2dA93R16D2BVCTXV5I\n0UeeGqwEFmQqu5P0p9uS7INlfQ2chQv208dRo18=\n=niL1\n-----END PGP MESSAGE-----\n
pgp__list_2__map_enc        = -----BEGIN PGP MESSAGE-----\n\nwcFMA9AgVmznaOk7ARAAUXL2mQiEjmrrJHOffPVpzG0pCU1/k/GtGntjgcD3CtCG\nMBA3vvMQw5QsIQZzccHqWmEvHZuveY6jr17TdeWyvD/nlGVjXe+xOytDx8RBD/2P\nmhNDSb1PyXtPrbdsVvxEmhuiHUNiFgP0AHhiIvRrZfj3sh6fp/RrveQZCPew9H7S\nGoQm8sRtitSWYa9/qFCYB0wIsDreRP/Q9UppsnOUW9R+v/PxCzKe5dnJ9LBWYUgo\no6y+Pe3YI+jBomg65fkzDs1lcHhK/tQ3AdhxRwbDKSIWo/kzoNd5Jqtogbue0RAm\n9lhRAZM8Ngm6xRK/4RnAcQtS4W6aMLyzQtYGRSpmo9yJzeG9eUARtCjKJxyRdwHI\n5NNGc6WrQdm2jgYQfZGbWxjps99tU1lzXLKNn3KaFyp7m6+BooJyxImGpRBBm+/n\nLhxjjcZjLH9jp0VcGYUDoIpuNBlgvW5pgwkH2T1OC9SPCF5bLjFZE+MO6YOy/yqt\nxcbGlQ6u9zjpiC07QXB6bKJqW93GjXMSZTSDKbTCaSEMMf3iJOSlJC645EvDw3xy\nYjP9bZhs0a1BIkO6uXYi5Z63TOtGpxh6O/M2mL5lxemkq7iIe2k0eN4fGa5AWqvE\nnR6/TkodW9ssJxtOLWtoH2wKz+4EcE2E74o2AR8tLDtscJzq8Lvck+p/SGut1GLS\n4AHk/d8c+118TQqtdBtIZePUVOHDWeA+4GHhgvrgdeIFjiZI4GnlpsgtWMg2D4NT\nvLI0ghoUYkza8Z5BnXyaWURjna2ZV6fgVeTm4yi8iAgD6XRmkwHOtLZS4o+JLlbh\nWuIA\n=Sx2X\n-----END PGP MESSAGE-----
```

As you can see most of the settings is still readable and reviewable, only credential fields are encrypted.

Then we will use [kustomize-sopssecretgenerator](https://github.com/goabout/kustomize-sopssecretgenerator) plugin to Kustomize, to decrypt those files and generate secrets from it when building the manifests - in the same fashion, how [Kustomize secretGenerator](https://kubectl.docs.kubernetes.io/pages/reference/kustomize.html#secretgenerator) works. That way have maintain all of these aspects all together:
- Readability, clarity and revieability in VCS
- Deployable via Kustomize
- Sensitive data are encrypted
- Configuration file is extendable
- Solgate relies on a single config file = stable and predictable API

**Comment by tumido** [comment]
@durandom The "mixing of encrypted secret and plaintext config" would happen in VCS only. When deployed in OpenShift, it's all always unencrypted, base64-encoded Kubernetes secret resource. And since it would be all in a single secret, we can just simply mount it to the solgate task. :wink:

**Comment by tumido** [comment]
Let's rather demonstrate it all by an example. :wink:

Unfortunately we can't use environment variables for the secrets, since each pipeline instance may use different number of sync locations, therefore different amount of credentials.

I want to store all instance configuration in a single config file. This config is stored as a secret in OpenShift and [mounted as a file in runtime](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-files-from-a-pod) and looks like this:

```ini
; /etc/solgate.ini

[source_raw]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
formatter = {date}/{collection}.{ext}
base_path = DH-PLAYPEN/storage/input

[destination_unpack_historic]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
formatter = {collection}/historic/{date}-{collection}.{ext}
base_path = DH-PLAYPEN/storage/output
unpack = yes

[destination_unpack_latest]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
formatter = {collection}/latest/full_data.csv
base_path = DH-PLAYPEN/storage/output
unpack = yes

[destination_raw]
aws_access_key_id = KEY_ID
aws_secret_access_key = ACCESS_KEY
endpoint_url = https://s3.upshift.redhat.com
base_path = DH-PLAYPEN/storage/output

[solgate]
alerts_smtp_server = smtp.corp.redhat.com
alerts_from        = solgate-alerts@redhat.com
alerts_to          = dev-null@redhat.com
timedelta          = 1d
```

As you can see, this instance will sync from `source_raw` to 3 destinations `destination_*`

We want to track this INI file in GIT in as readible format as possible. So in GIT, we encrypt only `aws` specific keys of it via SOPS:
```yaml
# manifests/.sops.yaml

creation_rules:
  - encrypted_regex: "^(aws.*)$"
    pgp: "EFDB9AFBD18936D9AB6B2EECBD2C73FF891FBC7E, A76372D361282028A99F9A47590B857E0288997C, 04DAFCD9470A962A2F272984E5EB0DA32F3372AC"
```

And the resulting config file will look like this:

```ini
; manifests/overlays/dev/this_instance/config.ini
[source]
aws_access_key_id     = ENC[AES256_GCM,data:lpy0WhN4e+o5spYSyBr4G0Bp5W8=,iv:9G/7iI/E46vU5k+gxcMHW1mmArzeqHYr76a4Lg8q2/4=,tag:gAdwPCSx9axGwojRyEQQWA==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:PZ91CnbyWNHX9aTVSNR0HDWTTMPMtaGMxWGYshKH1M7Vo5xvCh1QHw==,iv:/+jUkv7TLRdrRVzu8DG8tC05snT5mlaT6XAq+lslS4k=,tag:e4SjZGSzlnm5LMyej9U5hQ==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
formatter             = {date}/{collection}.{ext}
base_path             = DH-PLAYPEN/storage/input

[destination_unpack_historic]
aws_access_key_id     = ENC[AES256_GCM,data:96ftEMbEpt4qAi/6LK4pzCyJLDM=,iv:U6dGaPBZXfee7XCj1ZhmGu75FJ9Z4mjDVtlVkKq4uDE=,tag:seihxPfpNB4EEbP1Wq3CDw==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:CGMHXL0abcviOXTe/gQ/dvegn2iXp3hwdPAYIXjnjMHq/Wcz0Kv43Q==,iv:IguBm4gWku76kRGtjrU7yb+LK749oWDh2R9sdMpytx8=,tag:JXqbKa/36ESV4tTkOFMYyQ==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
formatter             = {collection}/historic/{date}-{collection}.{ext}
base_path             = DH-PLAYPEN/storage/output
unpack                = yes

[destination_unpack_latest]
aws_access_key_id     = ENC[AES256_GCM,data:d3cntWCYKdjldnf+feRiNcFoL/s=,iv:SP158uN1Vat1P+RipnQBO8328Z6smuo91iPwco1720k=,tag:EWyNsngiv/sddqKAAk35nQ==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:8CzkDrXCKBHnkHA7iP4E7YsglBRd+kTRzSidBICgofPr6pu+eEYFIA==,iv:AEiL4jc0L7fsqfT/9vErsHSOkbAqjYuCFU05ngoKoNk=,tag:q6SnLrl5tiQUC2j0PBjOUg==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
formatter             = {collection}/latest/full_data.csv
base_path             = DH-PLAYPEN/storage/output
unpack                = yes

[destination_raw]
aws_access_key_id     = ENC[AES256_GCM,data:0eroBlaIxNzhCzbgf+VZensrlyA=,iv:yGiiLyIq4x2lhB7lMKuR0A/0btwQcuIDJTpnS80HD6Q=,tag:G4M+Lj5u0XqmMdCAaKJ4pA==,type:str]
aws_secret_access_key = ENC[AES256_GCM,data:ah+uxPlDQ0I6i7ZUagWpqno735GxEF8bqwKc2jZ0mNSVBuS9Xyj5uQ==,iv:myzLeh7goVyyoXdHP2IPvCcmsVo7wRJ+Bm3e56Apxk4=,tag:ne1oKcPVFK46PjS0x2KnKw==,type:str]
endpoint_url          = https://s3.upshift.redhat.com
base_path             = DH-PLAYPEN/storage/output

[solgate]
alerts_smtp_server = smtp.corp.redhat.com
alerts_from        = solgate-alerts@redhat.com
alerts_to          = dev-null@redhat.com
timedelta          = 1d

[sops]
pgp__list_0__map_fp         = EFDB9AFBD18936D9AB6B2EECBD2C73FF891FBC7E
pgp__list_2__map_created_at = 2020-08-07T10:29:29Z
pgp__list_2__map_fp         = 04DAFCD9470A962A2F272984E5EB0DA32F3372AC
lastmodified                = 2020-08-07T10:29:30Z
pgp__list_1__map_enc        = -----BEGIN PGP MESSAGE-----\n\nhQIMA5MpUvZdp4z9AQ//WgCo6KcapKGj1sQopw579+UyCFL0zeEQwVElsmKKVg5N\nvnncjEATrSMo5Ojvmi5JsXka9S8pCx19nB1TuwNTbeCHr901e0W6yy6Qvoe901Ei\npJtGgZChckPpg8JozDU1paYlbS/YWd5hDbKukd6BGSiKViHhAyWtYVg1A/4kFY06\nCflgJFwvmGTUf6W6zkOYjpfK/MvMYi/8zpSo5myRNf7MZJmDStFRlp/f2+SxrnCV\njf3Qy00a6mJtYDLvTDlydhdjLXHA7J4NARah7lXqHM2s5gmeUtUB1DNtB9q3Th3N\npsfM+5xwOacATX7eIVxJ5PbjZ+c+ODVBH2Jzysw5YDvdnr4dGsUzngvAx0JqLCxf\nHy/g8Kxd8ZWx0yMyIY3Kfvi4a+lvZ+wiiW8CcLQCLWmcuQJ8dFWm3/P03Hf9Jgbj\n2K8hlf3PhBr7mgnzG5h9mYIHBRSwFYMZ05GgKn5mMMwtp6FvFrY5xRPO1yuGWRPJ\nKYts34EKITRZnvQB8/nK/V3H6C7mVFYGIvTFysH/XNS/zoVLNEfuKvxiSSsvGZM2\nCuLZBAgz7TMUg0x5Di3VCMCaSqYAaYwDqGoeQAZgdD+fvknhzdXdu45NwoabTCNZ\nOdjENyjrAUBwZFvO4w6qK1NWswcl2qdXMtxryE9s1kvpd8e+UaE0yom+a5v5LW/S\nXAGuE/4FhljCA/3IdLrKHV3NTMlMo4CNgLQHBdfBW5jR0NbSeyUtYczOsme/+c9J\n5mMOW9NW4WvBODEEjIj95AvjLWNP4s/FR0u6gQ925ZQhl0vCfbakY0bfnMaD\n=uaW0\n-----END PGP MESSAGE-----\n
mac                         = ENC[AES256_GCM,data:oKrctMbVjP4s8natZaKyMgERF58xzY3ANuWFZoJ+FMQHa9wdX+8auWAkG3T3+1hZ1Si3rzso18yAmWnu5XJJ5MHeFv151mYMfmPGuD/umxA1rs1O86AMxMD+F4WOrJV8Q+Oq6CvkUnvimoF43bkV5y5kEEtOP4m1Fo42bywBkyI=,iv:FwiWswP0czEObEZBXLYXrx3zY4sf32hktQChMR7mLyk=,tag:wb7EcS/BASG5MIZBwz/lOw==,type:str]
version                     = 3.6.0
pgp__list_0__map_created_at = 2020-08-07T10:29:29Z
pgp__list_1__map_created_at = 2020-08-07T10:29:29Z
pgp__list_1__map_fp         = A76372D361282028A99F9A47590B857E0288997C
encrypted_regex             = ^(aws.*)$
pgp__list_0__map_enc        = -----BEGIN PGP MESSAGE-----\n\nhQEMA/irrHa183bxAQf9HR9cJM+dhrmO5VkVwcj/Lh3+/CcclK0SDFVP3Hz0swH5\nUv04lQaoa9n/yap/633J61HnlWpitWE3HKnI+56fCaQb5nSZcRVU82mIVpK2UN0y\n5UwJq8tg5DzkzjFgl6IolYCeEB4mvqBgUD04hf6M65H2tzuwOOvvrknv6gUekorJ\nN4VRwzUZRZwiYFxhLRj/kcWxCd+bdOGge4EfkN2RoVNRFXkQ0CNawEzxg3e6lFk2\n6ouMs1Geu2aZs/LO0sNliNidW9p6uUUFNuJ7GWdfIjwOTmZSVIBzPBpU1nvsdyhB\npZBdRJHUoYt3mu1uxkf9J9hKfcBqcbulUXDS9qHxrtJcASZAlAMTONjmeZcVcPf5\n1RGT0uMp3L6lS22rt7FyAFnbaDJ1h8iUuPpHCX+8T67rXX2dA93R16D2BVCTXV5I\n0UeeGqwEFmQqu5P0p9uS7INlfQ2chQv208dRo18=\n=niL1\n-----END PGP MESSAGE-----\n
pgp__list_2__map_enc        = -----BEGIN PGP MESSAGE-----\n\nwcFMA9AgVmznaOk7ARAAUXL2mQiEjmrrJHOffPVpzG0pCU1/k/GtGntjgcD3CtCG\nMBA3vvMQw5QsIQZzccHqWmEvHZuveY6jr17TdeWyvD/nlGVjXe+xOytDx8RBD/2P\nmhNDSb1PyXtPrbdsVvxEmhuiHUNiFgP0AHhiIvRrZfj3sh6fp/RrveQZCPew9H7S\nGoQm8sRtitSWYa9/qFCYB0wIsDreRP/Q9UppsnOUW9R+v/PxCzKe5dnJ9LBWYUgo\no6y+Pe3YI+jBomg65fkzDs1lcHhK/tQ3AdhxRwbDKSIWo/kzoNd5Jqtogbue0RAm\n9lhRAZM8Ngm6xRK/4RnAcQtS4W6aMLyzQtYGRSpmo9yJzeG9eUARtCjKJxyRdwHI\n5NNGc6WrQdm2jgYQfZGbWxjps99tU1lzXLKNn3KaFyp7m6+BooJyxImGpRBBm+/n\nLhxjjcZjLH9jp0VcGYUDoIpuNBlgvW5pgwkH2T1OC9SPCF5bLjFZE+MO6YOy/yqt\nxcbGlQ6u9zjpiC07QXB6bKJqW93GjXMSZTSDKbTCaSEMMf3iJOSlJC645EvDw3xy\nYjP9bZhs0a1BIkO6uXYi5Z63TOtGpxh6O/M2mL5lxemkq7iIe2k0eN4fGa5AWqvE\nnR6/TkodW9ssJxtOLWtoH2wKz+4EcE2E74o2AR8tLDtscJzq8Lvck+p/SGut1GLS\n4AHk/d8c+118TQqtdBtIZePUVOHDWeA+4GHhgvrgdeIFjiZI4GnlpsgtWMg2D4NT\nvLI0ghoUYkza8Z5BnXyaWURjna2ZV6fgVeTm4yi8iAgD6XRmkwHOtLZS4o+JLlbh\nWuIA\n=Sx2X\n-----END PGP MESSAGE-----
```

As you can see most of the settings is still readable and reviewable, only credential fields are encrypted.

Then we will use [kustomize-sopssecretgenerator](https://github.com/goabout/kustomize-sopssecretgenerator) plugin to Kustomize, to decrypt those files and generate secrets from it when building the manifests - in the same fashion, how [Kustomize secretGenerator](https://kubectl.docs.kubernetes.io/pages/reference/kustomize.html#secretgenerator) works. That way have maintain all of these aspects all together:
- Readability, clarity and revieability in VCS
- Deployable via Kustomize
- Sensitive data are encrypted
- Configuration file is extendable
- Solgate relies on a single config file = stable and predictable API

**Comment by tumido** [comment]
Solved via #20
